Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm

```diff
@@ -1159,13 +1159,87 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+    if (!currentPage)
+        return nullptr;
+
+    RetainPtr<NSArray> annotationsForCurrentPage = [currentPage annotations];
+    auto indexOfCurrentAnnotation = [annotationsForCurrentPage indexOfObject:currentAnnotation.get()];
+    ASSERT(indexOfCurrentAnnotation != NSNotFound);
+    if (indexOfCurrentAnnotation == NSNotFound)
+        return nullptr;
+
+    auto isForwardSearchDirection = searchDirection == AnnotationSearchDirection::Forward;
+    if ((isForwardSearchDirection && indexOfCurrentAnnotation + 1 < [annotationsForCurrentPage count]) || (!isForwardSearchDirection && indexOfCurrentAnnotation)) {
+        auto startingIndexForSearch = isForwardSearchDirection ? indexOfCurrentAnnotation + 1 : indexOfCurrentAnnotation - 1;
+        if (RetainPtr<PDFAnnotation> nextTextAnnotationOnCurrentPage = getFirstTextAnnotationStartingAtIndex(annotationsForCurrentPage, startingIndexForSearch, searchDirection))
+            return nextTextAnnotationOnCurrentPage;
+    }
+
+    auto indexForCurrentPage = m_documentLayout.indexForPage(currentPage);
+    if (!indexForCurrentPage)
+        return nullptr;
+
+    RetainPtr<PDFAnnotation> nextAnnotation = nullptr;
+    auto nextPageToSearchIndex = indexForCurrentPage.value();
+    while (!nextAnnotation) {
+        auto getNextPageToSearchIndex = [this, isForwardSearchDirection](unsigned currentPageIndex) -> unsigned {
+            auto pageCount = m_documentLayout.pageCount();
+            if (!isForwardSearchDirection && !currentPageIndex)
+                return pageCount - 1;
+            return isForwardSearchDirection ? ((currentPageIndex + 1) % pageCount) : currentPageIndex - 1;
+        };
+        nextPageToSearchIndex = getNextPageToSearchIndex(nextPageToSearchIndex);
+        RetainPtr<PDFPage> nextPage = m_documentLayout.pageAtIndex(nextPageToSearchIndex);
+        if (!nextPage)
+            continue;
+        if (RetainPtr<NSArray> nextPageAnnotations = [nextPage annotations]; nextPageAnnotations && [nextPageAnnotations count])
+            nextAnnotation = getFirstTextAnnotationStartingAtIndex(nextPageAnnotations, isForwardSearchDirection ? 0 : [nextPageAnnotations count] - 1, searchDirection);
+    }
+    return nextAnnotation;
+}
 
 void UnifiedPDFPlugin::focusNextAnnotation()
 {
+    if (!m_activeAnnotation)
+        return;
+    RetainPtr<PDFAnnotation> nextTextAnnotation = getNextTextAnnotation(AnnotationSearchDirection::Forward);
+    if (!nextTextAnnotation || nextTextAnnotation == m_activeAnnotation->annotation())
+        return;
+    setActiveAnnotation(nextTextAnnotation.get());
 }
 
 void UnifiedPDFPlugin::focusPreviousAnnotation()
 {
+    if (!m_activeAnnotation)
+        return;
+    RetainPtr<PDFAnnotation> previousTextAnnotation = getNextTextAnnotation(AnnotationSearchDirection::Backward);
+    if (!previousTextAnnotation || previousTextAnnotation == m_activeAnnotation->annotation())
+        return;
+    setActiveAnnotation(previousTextAnnotation.get());
 }
```

Comment:
Should we be focusing on the first / last annotation (respectively) if there is no active annotation?

Reply:
I don't think so. The current behavior seems to be that the focusing keyboard shortcuts do not do anything if an annotation is currently not active